pkg/scheduler/plugins/elasticquota/core/group_quota_manager.go

```diff
@@ -0,0 +1,616 @@
+	totalResourceIndeed v1.ResourceList // totalResource without systemQuotaGroup and DefaultQuotaGroup's used Quota
+	totalResource       v1.ResourceList // totalResource with systemQuotaGroup and DefaultQuotaGroup's used Quota
+	resourceKeys        map[v1.ResourceName]struct{}
+	// quotaInfoMap stores all the nodes, it can help get all parents conveniently
+	quotaInfoMap map[string]*QuotaInfo
+	// quotaTreeWrapperMap helps calculate the subGroups' runtimeQuota in one quotaGroup
+	quotaTreeWrapperMap map[string]*QuotaTreeWrapper
+	// quotaTopoNode only stores the topology of the quota
+	quotaTopoNode                    map[string]*QuotaTopoNode
+	scaleMinQuotaEnabled             bool
+	scaleMinQuotaWhenOverRootResInfo *ScaleMinQuotaWhenOverRootResInfo
+	once                             sync.Once
+}
+
+func NewGroupQuotaManager(systemGroupMax, defaultGroupMax v1.ResourceList) *GroupQuotaManager {
+	quotaManager := &GroupQuotaManager{
+		totalResourceIndeed:              v1.ResourceList{},
+		totalResource:                    v1.ResourceList{},
+		resourceKeys:                     make(map[v1.ResourceName]struct{}),
+		quotaInfoMap:                     make(map[string]*QuotaInfo),
+		quotaTreeWrapperMap:              make(map[string]*QuotaTreeWrapper),
+		quotaTopoNode:                    make(map[string]*QuotaTopoNode),
+		scaleMinQuotaWhenOverRootResInfo: NewScaleMinQuotaWhenOverRootResInfo(),
+	}
+	quotaManager.quotaInfoMap[extension.SystemQuotaName] = NewQuotaInfo(true, true, extension.SystemQuotaName, extension.SystemQuotaName)
+	quotaManager.quotaInfoMap[extension.SystemQuotaName].setMaxQuotaNoLock(systemGroupMax)
+	quotaManager.quotaInfoMap[extension.DefaultQuotaName] = NewQuotaInfo(true, true, extension.DefaultQuotaName, extension.DefaultQuotaName)
+	quotaManager.quotaInfoMap[extension.DefaultQuotaName].setMaxQuotaNoLock(defaultGroupMax)
+	quotaManager.quotaTreeWrapperMap[extension.RootQuotaName] = NewQuotaTreeWrapper(extension.RootQuotaName)
+	return quotaManager
+}
+
+func (gqm *GroupQuotaManager) SetScaleMinQuotaEnabled(flag bool) {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	gqm.scaleMinQuotaEnabled = flag
+	klog.V(3).Infof("Set ScaleMinQuotaEnabled, flag:%v", gqm.scaleMinQuotaEnabled)
+}
+
+func (gqm *GroupQuotaManager) UpdateClusterTotalResource(deltaRes v1.ResourceList) {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	klog.V(3).Infof("UpdateClusterResource deltaRes:%v", deltaRes)
+	gqm.updateClusterTotalResourceNoLock(deltaRes)
+}
+
+func (gqm *GroupQuotaManager) updateClusterTotalResourceNoLock(deltaRes v1.ResourceList) {
+	gqm.totalResource = quotav1.Add(gqm.totalResource, deltaRes)
+
+	sysAndDefaultUsed := quotav1.Add(gqm.quotaInfoMap[extension.DefaultQuotaName].GetUsed(), gqm.quotaInfoMap[extension.SystemQuotaName].GetUsed())
+	totalResNoSysOrDefault := quotav1.Subtract(gqm.totalResource, sysAndDefaultUsed)
+
+	diffRes := quotav1.Subtract(totalResNoSysOrDefault, gqm.totalResourceIndeed)
+
+	if !quotav1.IsZero(diffRes) {
```

Comment:
the diff check can put it in runtimeQuotaCalculator's SetClusterTotalResource if we kill gqm.totalResourceIndeed. and log can be judge by SetClusterTotalResource's return value.